components/util/model.jsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { immutableMerge } from './index.js';
+
+export function Model({ initialState, methods }) {
+  const listeners = [];
+  const notifyListeners = nextState => listeners.forEach(fn => fn(nextState));
+
+  const model = {
+    state: Object.freeze(initialState),
+    subscribe: fn => listeners.push(fn),
+    unsubscribe: fn => listeners.splice(listeners.indexOf(fn), 1),
+  };
+
+  Object.keys(methods).forEach((method) => {
+    model[method] = (...args) => {
+      const updates = methods[method].apply(model, args);
+      if (updates && typeof updates === 'object') {
+        model.state = immutableMerge(model.state, updates);
+        notifyListeners(model.state);
+      }
+      // could also check here to see if fn returns a promise and then wait
+      // until it resolves to merge and notifyListeners. that would make
+      // async methods really nice to work with.
+      // ...should this return  something? maybe return `updates`?
```

Comment:
I like this idea a lot. Do you have an example use case? Does it seem like it would introduce a lot of edge cases or anything?